README.md
# LANTIS EasyLink Router 3
LANTIS is a system that allows you use simple low end Linux systems (Raspberry Pi and bellow) to securely route ports to the internet and bypass the inbound NAT rule requirement. All using SSH as the secure transport. 

## Features
* Easy CSV File to manage mappings (with inheritance)
* Automatic Connection Recovery and Internet Connection Verification
* Instant Public IP Address Update (when using Direct Connection Mode)
* Inbound NAT Bypass (uses Reverse Tunneling to bypass NAT)
* RSA Encrypted Data Transport
* Port Hijack (Kills whatever is using the port and hijacks access to port)
* Linked Connection Mode (Multiple Ports on a single connections)
* Automatic Setup and Key Exchange (See Setup Guide)
* Minimal System Footprint
* Leaves little evidence (beware of SSH logs)
* Allows usage of Multiple Remote and Local Hosts for routing

## Example use case
* Privacy of your location
* Changing dynamic IP
* No control over the router
* Can’t open ports due to security
* Double-NAT (Behind multiple NAT routers)

## Requirements
* Remote Linux based Hosts (3 if you want to run watchdog separately) to present port
* Local Linux based Host to create connection + watchdog (Watchdog can be ran from a separate host)
* Root SSH Access Recommended to Remote Host
* Trusted SSH Key for Setup
* Outbound SSH
* BASH and openSSH

## Setup
1 Install a Raspberry Pi or any other system
2 Generate a SSH Key pair
3 Clone this Git
4 cd LANTIS/
5 Generate a SSH Key pair with the name lantis.key
6 Get a remote host from Digital Ocean or Vultr (Cheaper) with lowest hardware
7 Add your user SSH key to host for setup (Confirm you can login)
8 Configure CSV File and config files
9 Run with -X option in front for Dry Run (Connects and does setup but does not route packets) 

## Configuration

### Connection Table (ports.lantis.csv)
Example File:
```e;web;25.97.71.244;22;root;~;65500;root;0;0;0;80;127.0.0.1;80;1;
e;web;^;^;^;^;^;^;^;^;^;443;^;443;^;
l;admin;45.76.9.245;^;^;^;^;^;^;^;^;9998;192.168.0.90;3389;^;
^;9999;192.168.0.92;22;^;
>;6100;192.168.0.91;^;^;
e;rdp-1;^;^;^;127.0.0.1;65401;^;^;1;^;4000;192.168.0.101;3389;^;
e;rdp-2;^;^;^;^;65402;^;^;^;^;4001;192.168.0.102;^;^;
e;rdp-3;^;^;^;^;65403;^;^;^;^;4002;192.168.0.103;^;^;
d;rdp-4;^;^;^;^;65404;^;^;^;^;4003;192.168.0.104;^;^;
```

Web Server
* Sets Initial Settings
* Direct Connection with Dynamic IP
* Forward Local Host Port 80 and 443

```
;web;25.97.71.244;22;root;~;65500;root;0;0;0;80;127.0.0.1;80;1;
e;web;^;^;^;^;^;^;^;^;^;443;^;443;^;
```

Linked Connection
* Another end point but keep all other Settings
* 3 ports using same connection

```
l;admin;45.76.9.245;^;^;^;^;^;^;^;^;9998;192.168.0.90;3389;^;
^;9999;192.168.0.92;22;^;
>;6100;192.168.0.91;^;^;
```

Bypass NAT
* Bypass NAT Enabled
* Using Local Host as transport (Don't have to but simplest)
* 4 RDP Ports Open

```
e;rdp-1;^;^;^;127.0.0.1;65401;^;^;1;^;4000;192.168.0.101;3389;^;
e;rdp-2;^;^;^;^;65402;^;^;^;^;4001;192.168.0.102;^;^;
e;rdp-3;^;^;^;^;65403;^;^;^;^;4002;192.168.0.103;^;^;
d;rdp-4;^;^;^;^;65404;^;^;^;^;4003;192.168.0.104;^;^;
```

#### Options
1 Status (Enable [e], Disable [d], Linked [l])
2 Connection Name
3 Remote Host IP or FQDN
4 Remote Host SSH Port
5 Remote Host Login User
6 Local Host IP or FQDN
7 Local Host SSH Port
8 Local Host Login User
9 Setup Mode (1 = Setup Remote Host)
10 Bypass NAT (aka Reverse Tunneling)
11 Port Hijack
12 Public Port
13 Local Server IP or FQDN
14 Local Server Port
15 Listen Publicly or Local to Remote Host

#### Special Options
* ( ~ )(6: Local Host IP or FQDN) Dynamic Public IP (Only Direct Mode)
* ( ^ )(All but 1,2,12) Use last value
* ( l )(1: Status) Linked Connection, Use one connection for multiple ports
..* Use `^;5999;192.168.0.47;22;^;` for other ports (12,13,14,15)
..* Use ( > ) for last port (Used as end stop, without this the whole connection will be ignored)

#### Notes
* You can use multiple connections with the same name and they will be parsed as the connection group
* To use Bypass NAT you must set 6 to 127.0.0.1, 10 to 1, and 7 to a unique port for each connection
..* Does not yet support Linked Bypass mode (Multiple Connection trough the same tunnel), use normal linked mode
* In linked mode, if one port were to get overloaded and cause a buffer overflow the whole link will drop will all linked ports and will have to automatically reconnect. (Just a word of warning)
* Disabled Connection are still read for inheritance

### Config Files
Defaults are fine unless you have a need to change some of the values

#### lantis.config
```
# Log File used for watchdogs
LOG_FILE="./lantis.log"
# Time to pause after launching a connection
TIME_LAUNCH_PAUSE=4
# Time to pause after dropping a connection
TIME_DROP_PAUSE=2
```

#### watchdog.lantis.config
```
# Time to pause between a connection drop
TIME_FAILED_CONN=2
# Time to paues between a internet connection drop
TIME_FAILED_INET=5
# Timeout for internet test
TIMEOUT_VERIFY_INET=15
# Host to use for internet verification
HOST_VERIFY="https://google.com"
# SSH Command
CMD_SSH="ssh"
# SCP Command
CMD_SCP="scp"
# LANTIS Key used for 2-way auth
KEY=lantis.key
# Setup key used for setting up servers
SETUP_KEY="$HOME/.ssh/id_rsa"
# Common SSH options
COMMON_OPT="-C -2 -o BatchMode=yes -o StrictHostKeyChecking=no -o TCPKeepAlive=yes -o ServerAliveInterval=5 -o ConnectTimeout=15 -o LogLevel=Error"
# Endpoint Connection Back SSH Options
LOCAL_OPT="-N -o CompressionLevel=9 -o ExitOnForwardFailure=yes"
```

## Usage
```
root@lantis-controller:~/LANTIS# bash lantis.bash
= LANTIS Router 3 - Academy City Research =========
[---------][Mon  3 Jul 17:04:16 UTC 2017][ OK ] System Ready
 support : help.lantis.project@acr.moe

 LANTIS EasyLink Router 2 - Usage:

 Launch ================================================================
 -l  Launch a Connecion           -L  Launch ALL Connections

 Drop ==================================================================
 -k  Drop a Connection            -K  Drop ALL Connections

 Extra Options (MUST be set before a action) ===========================
 -C  Use another port list (Default: ./ports.lantis.csv)
 -X  Dry Run (Does not do any action but test full connection)
 -Z  Display Setup Guide

 NOTE: You can give multiple actions (ex: -l admin -k ssh -l rdp)

root@lantis-controller:~/LANTIS#
```
### Launching a connection
```
root@lantis-controller:~/LANTIS# bash lantis.bash -C production.csv -l kazari-1
= LANTIS Router 3 - Academy City Research =========
[---------][Mon  3 Jul 17:05:44 UTC 2017][ OK ] System Ready
[kazari-1][Mon  3 Jul 17:05:47 UTC 2017][INFO] Launching Connection...
root@lantis-controller:~/LANTIS#
```
#### Log Output
```
[kazari-1][Mon  3 Jul 17:05:48 UTC 2017][INFO] DATA LOADED
[kazari-1][Mon  3 Jul 17:05:49 UTC 2017][INFO] Outbound Internet Connection: OK
[kazari-1][Mon  3 Jul 17:05:49 UTC 2017][INFO] Outbound End-Point:           OK
[kazari-1][Mon  3 Jul 17:05:51 UTC 2017][INFO][>>>] Establishing Control...
[kazari-1][Mon  3 Jul 17:05:51 UTC 2017][INFO][<<<] Linked!
```

CHANGELOG
Nothing yet sense im to lazy to edit this

lantis.bash
#!/bin/bash

# LANTIS EasyLink 2 #
if [ -f ./.stat.inet ]; then rm -f ./.stat.inet; echo "[---------][$(date "${DATE_FORMAT}")][WARN] Watchdog State Control will be dropped"; fi
# OPERATIONS ###########################################################################################################
HEADER() {
cat << EOF
     __    ___    _   _________________
    / /   /   |  / | / /_  __/  _/ ___/
   / /   / /| | /  |/ / / /  / / \__ \ 
  / /___/ ___ |/ /|  / / / _/ / ___/ / 
 /_____/_/  |_/_/ |_/ /_/ /___//____/  
 Lain Anonymous NetworkIng System
 by : Academy City Research

EOF
}
USAGE(){
cat << EOF
 support : help.lantis.project@acr.moe

 LANTIS EasyLink Router 2 - Usage:
 
 Launch ================================================================
 -l  Launch a Connecion           -L  Launch ALL Connections
 
 Drop ==================================================================
 -k  Drop a Connection            -K  Drop ALL Connections

 Extra Options (MUST be set before a action) ===========================
 -C  Use another port list (Default: ./ports.lantis.csv)
 -X  Dry Run (Does not do any action but test full connection)
 -Z  Display Setup Guide

 NOTE: You can give multiple actions (ex: -l admin -k ssh -l rdp)

EOF
}
if [ $# -lt 1 ]; then HEADER; USAGE; exit 0; fi
SETUPGUIDE(){
cat << EOF
 support : help.lantis.project@acr.moe

 Setup Guide:

 1) Generate a SSH key in ~/.ssh/id_rsa
      [root@local]~ ssh-keygen –b 4086
 2) Add id_rsa.pub to /root /.ssh/authorized_keys on the remote server
 3) Generate a SSH key for LANTIS in your current directory where the script will stay
      [root@local]~/LANTIS ssh-keygen –b 4096 –f lantis.key
 4) Add the entry to the port list with the setup bit set to 1, run that connection
 5) You should be fine as long as your destination host allows your default ssh key

 CSV File Guide:
 
 NOTE: You can use "^" to use last value
 NOTE: You can use "~" for Local Host to use your current Public IP (Updated on Connection Drop)
 NOTE: Disabled connections are still read
 NOTE: You can only use Public Ports 1-2014 if you are using Remote User root
 WARN: There is little input verification, make sure you know what your doing
 WARN: You cannot use the same local port when using Bypass NAT
 
 1 ) Enable or Disale Connection [e,d]
 2 ) Connection Name [string.no spaces]
 3 ) Remote Host [string]
 4 ) Remote Port [int]
 5 ) Remote User [string]
 6 ) Local Host [~ for dynamic ip]
 7 ) Local Port [int]
 8 ) Local User [string]
 9 ) Setup End Point [y or n]
 10) Bypass Inbound NAT [y or n]
 11) Hijack Port [y or n]
 12) Public Server Port [int]
 13) Local Server Host [string]
 14) Local Server Port [int]
 15) Public or Remote Host-Only [p or l]
 
EOF
}
FORKER () {
echo "[${CONNECTION_NAME}][$(date "${DATE_FORMAT}")][INFO] $(if [ ${1} = 1 ]; then echo "Launching"; elif [ ${1} = 2 ]; then echo "Dropping"; fi) Connection..."
if [ ${DRY} -eq 1 ]; then 
	echo "./.watchdog.lantis.bash -n ${CONNECTION_NAME} -h ${REMOTE_HOST} -p ${REMOTE_PORT} -u ${REMOTE_USER} \
	-H ${LOCAL_HOST} -P ${LOCAL_PORT} -U ${LOCAL_USER} ${PORT_FWDLN}${EXTRA_OPT}"
	bash ./.watchdog.lantis.bash -n ${CONNECTION_NAME} -h ${REMOTE_HOST} -p ${REMOTE_PORT} -u ${REMOTE_USER} \
	-H ${LOCAL_HOST} -P ${LOCAL_PORT} -U ${LOCAL_USER} ${PORT_FWDLN}${EXTRA_OPT} -m ${1} -X ${DRY}
else 
	pkill -f "^bash ./.watchdog.lantis.bash -n ${CONNECTION_NAME} *." > /dev/null
	nohup bash ./.watchdog.lantis.bash -n ${CONNECTION_NAME} -h ${REMOTE_HOST} -p ${REMOTE_PORT} -u ${REMOTE_USER} \
	-H ${LOCAL_HOST} -P ${LOCAL_PORT} -U ${LOCAL_USER} ${PORT_FWDLN}${EXTRA_OPT}-m ${1} &>> ${LOG_FILE} & 
fi
sleep $(if [ ${1} = 1 ]; then echo "${TIME_LAUNCH_PAUSE}"; elif [ ${1} = 2 ]; then echo "${TIME_DROP_PAUSE}"; fi)
}
WATCHDOG() {
while read in; do
if [[ $(echo $in | awk -F '[ ]' '{print $1}') != "#" ]]; then
	if [[ $(echo $in | awk -F '[;]' '{print $1}') = "e" ]] || [[ $(echo $in | awk -F '[;]' '{print $1}') = "d" ]] || [[ $(echo $in | awk -F '[;]' '{print $1}') = "l" ]]; then
		SKIP=0; EXTRA_OPT=""; 
		CONNECTION_STATUS=$(echo $in | awk -F '[;]' '{print $1}')  #Enabled[E or D]
		CONNECTION_NAME=$(echo $in | awk -F '[;]' '{print $2}')    #Name[string]
		L_REMOTE_HOST=$(echo $in | awk -F '[;]' '{print $3}')      #Remote Host[string]
		L_REMOTE_PORT=$(echo $in | awk -F '[;]' '{print $4}')      #Remote Port[string]
		L_REMOTE_USER=$(echo $in | awk -F '[;]' '{print $5}')      #Remote User[string]
		L_LOCAL_HOST=$(echo $in | awk -F '[;]' '{print $6}')       #Local Host[string]
		L_LOCAL_PORT=$(echo $in | awk -F '[;]' '{print $7}')       #Local Port[string]
		L_LOCAL_USER=$(echo $in | awk -F '[;]' '{print $8}')       #Local User[string]
		L_REMOTE_SETUP=$(echo $in | awk -F '[;]' '{print $9}')     #Remote Setup[strng]
		L_LOCAL_OPEN=$(echo $in | awk -F '[;]' '{print $10}')      #Bypass NAT[int]
		L_REMOTE_KILL=$(echo $in | awk -F '[;]' '{print $11}')     #Kill Access[int]
		REMOTE_FWDPORT=$(echo $in | awk -F '[;]' '{print $12}')    #Server Remote Port[int]
		L_LOCAL_FWDHOST=$(echo $in | awk -F '[;]' '{print $13}')   #Server Local Host[string]
		L_LOCAL_FWDPORT=$(echo $in | awk -F '[;]' '{print $14}')   #Server Local Port[int]
		L_REMOTE_FWDPUB=$(echo $in | awk -F '[;]' '{print $15}')   #Public or Local[P or L]
		# Read Value in Memory if "^"
		if [ ${L_REMOTE_HOST}   != "^" ]; then REMOTE_HOST=${L_REMOTE_HOST};     fi
		if [ ${L_REMOTE_PORT}   != "^" ]; then REMOTE_PORT=${L_REMOTE_PORT};     fi
		if [ ${L_REMOTE_USER}   != "^" ]; then REMOTE_USER=${L_REMOTE_USER};     fi
		if [ ${L_LOCAL_HOST}    != "^" ]; then LOCAL_HOST=${L_LOCAL_HOST};       fi
		if [ ${L_LOCAL_PORT}    != "^" ]; then LOCAL_PORT=${L_LOCAL_PORT};       fi
		if [ ${L_LOCAL_USER}    != "^" ]; then LOCAL_USER=${L_LOCAL_USER};       fi
		if [ ${L_REMOTE_SETUP}  != "^" ]; then REMOTE_SETUP=${L_REMOTE_SETUP};   fi
		if [ ${L_LOCAL_OPEN}    != "^" ]; then LOCAL_OPEN=${L_LOCAL_OPEN};       fi
		if [ ${L_REMOTE_KILL}   != "^" ]; then REMOTE_KILL=${L_REMOTE_KILL};     fi
		if [ ${L_LOCAL_FWDHOST} != "^" ]; then LOCAL_FWDHOST=${L_LOCAL_FWDHOST}; fi
		if [ ${L_LOCAL_FWDPORT} != "^" ]; then LOCAL_FWDPORT=${L_LOCAL_FWDPORT}; fi
		if [ ${L_REMOTE_FWDPUB} != "^" ]; then REMOTE_FWDPUB=${L_REMOTE_FWDPUB}; fi
		# Option Selector
		if [ ${REMOTE_FWDPUB} -eq 1 ]; then EXTRA_OPT="${EXTRA_OPT}-L "; fi
		if [ ${REMOTE_SETUP} -eq 1 ]; then EXTRA_OPT="${EXTRA_OPT}-S "; fi
		if [ ${LOCAL_OPEN} -eq 1 ]; then EXTRA_OPT="${EXTRA_OPT}-R "; fi
		if [ ${REMOTE_KILL} -eq 1 ]; then EXTRA_OPT="${EXTRA_OPT}-K "; fi
		
		PORT_FWDLN="-D ${REMOTE_FWDPORT}:${LOCAL_FWDHOST}:${LOCAL_FWDPORT} "
	elif [[ $(echo $in | awk -F '[;]' '{print $1}') = "^" ]] || [[ $(echo $in | awk -F '[;]' '{print $1}') = ">" ]]; then
		CONNECTION_STATUS=$(echo $in | awk -F '[;]' '{print $1}') #Enabled[E or D]
		REMOTE_FWDPORT=$(echo $in | awk -F '[;]' '{print $2}')    #Server Remote Port[int]
		L_LOCAL_FWDHOST=$(echo $in | awk -F '[;]' '{print $3}')   #Server Local Host[string]
		L_LOCAL_FWDPORT=$(echo $in | awk -F '[;]' '{print $4}')   #Server Local Port[int]
		L_REMOTE_FWDPUB=$(echo $in | awk -F '[;]' '{print $5}')   #Public or Local[P or L]
		
		if [ ${L_LOCAL_FWDHOST} != "^" ]; then LOCAL_FWDHOST=${L_LOCAL_FWDHOST}; fi
		if [ ${L_LOCAL_FWDPORT} != "^" ]; then LOCAL_FWDPORT=${L_LOCAL_FWDPORT}; fi
		if [ ${L_REMOTE_FWDPUB} != "^" ]; then REMOTE_FWDPUB=${L_REMOTE_FWDPUB}; fi
		
		PORT_FWDLN="${PORT_FWDLN}-D ${REMOTE_FWDPORT}:${LOCAL_FWDHOST}:${LOCAL_FWDPORT} "
	fi
	# Run Selector
	if   [ ${1} -eq 1 ] && [ ${#} -eq 1 ] && ( [ ${CONNECTION_STATUS} = "e" ] || [ ${CONNECTION_STATUS} = ">" ] ) ; then FORKER ${1}
	elif [ ${1} -eq 1 ] && [ ${#} -eq 2 ] && ( [ ${CONNECTION_STATUS} = "e" ] || [ ${CONNECTION_STATUS} = ">" ] ) && [ ${CONNECTION_NAME} = "${2}" ]; then FORKER ${1}
	
	elif [ ${1} -eq 2 ] && [ ${#} -eq 1 ] && ( [ ${CONNECTION_STATUS} = "e" ] || [ ${CONNECTION_STATUS} = ">" ] ); then FORKER ${1}
	elif [ ${1} -eq 2 ] && [ ${#} -eq 2 ] && ( [ ${CONNECTION_STATUS} = "e" ] || [ ${CONNECTION_STATUS} = ">" ] ) && [ ${CONNECTION_NAME} = "${2}" ]; then FORKER ${1}; fi
fi
done < $PORT_LIST
}
# SET VARS #############################################################################################################
DRY=0; PORT_LIST="./ports.lantis.csv"; LOG_FILE="./lantis.log"; TIME_LAUNCH_PAUSE=4; TIME_DROP_PAUSE=2; DATE_FORMAT='+%d/%m/%Y %H:%M:%S'
source ./.lantis.config
# MAIN RUNTIME #########################################################################################################
echo "= LANTIS Router 3 - Academy City Research ========="
echo "[---------][$(date "${DATE_FORMAT}")][ OK ] System Ready"
# PARSE INPUT ##########################################################################################################
while getopts "C:XLl:Kk:Z" opt; do 
  case $opt in
  	C) PORT_LIST="${OPTARG}";;
	X) DRY=1;;
	L) HEADER; WATCHDOG 1;;
	l) WATCHDOG 1 ${OPTARG};;
	K) HEADER; WATCHDOG 2;;
	k) WATCHDOG 2 ${OPTARG};;
    Z) HEADER; SETUPGUIDE; exit 0;;
    \?) echo "[PEBKAC] WTF is -$OPTARG?, thats not a accepted option, Abort"; USAGE; exit 1;;
    :) echo "[PEBKAC] -$OPTARG requires an argument, Abort"; USAGE; exit 1;;
  esac
done

watchdog.lantis.bash
#!/bin/bash

# LANTIS EasyLink 2 #
# OPERATIONS ###########################################################################################################
TEST_HOST_VERIFY () { 
${CMD_SSH} ${REMOTE_HOST} -l ${REMOTE_USER} -p ${REMOTE_PORT} -i ${KEY} ${COMMON_OPT} << EOF
echo "[${CONNECTION_NAME}][$(date)][INFO] Outbound End-Point:           OK"
EOF
}
TEST_HOST_FAILED () { 
echo "[${CONNECTION_NAME}][$(date)][ERR!] Outbound End-Point:    No Access"
if [ ${REMOTE_SETUP} -eq 1 ]; then
	echo "[${CONNECTION_NAME}][$(date)][INFO] Passing Key to End-Point..."
	${CMD_SCP} ${COMMON_OPT} -o Port=${REMOTE_PORT} ${KEY} ${REMOTE_USER}@${REMOTE_HOST}:${KEY}
	${CMD_SSH} ${REMOTE_HOST} -l ${REMOTE_USER} -p ${REMOTE_PORT} -i ${SETUP_KEY} ${COMMON_OPT} << EOF
		echo "$(cat ${KEY}.pub)" >> ~/.ssh/authorized_keys
EOF
	echo "[${CONNECTION_NAME}][$(date)][INFO] Passing Key to Local..."
	if [ ${LOCAL_IP} = "~" ] || [ ${LOCAL_OPEN} -eq 0 ]; then echo "$(cat ${KEY}.pub)" >> ~/.ssh/authorized_keys
	else 
		${CMD_SCP} ${COMMON_OPT} -o Port=${REMOTE_PORT} ${KEY} ${LOCAL_IP}@${LOCAL_USER}:${KEY}
		${CMD_SSH} ${REMOTE_HOST} -l ${REMOTE_USER} -p ${REMOTE_PORT} -i ${SETUP_KEY} ${COMMON_OPT} << EOF
		echo "$(cat ${KEY}.pub)" >> ~/.ssh/authorized_keys
EOF
	fi
	echo "[${CONNECTION_NAME}][$(date)][INFO] Key Exchange Complete"
fi
}
TEST_INET_VERIFY () {
wget -q --spider ${HOST_VERIFY} --timeout=${TIMEOUT_VERIFY_INET}
}
TEST_INET_PASSED () {
echo "[${CONNECTION_NAME}][$(date)][INFO] Outbound Internet Connection: OK"
}
TEST_INET_FAILED () {
echo "[${CONNECTION_NAME}][$(date)][ERR!] Outbound Internet Connection: Failed"
sleep ${TIME_FAILED_INET}
}
TEST_CONN_FAILED () {
echo "[${CONNECTION_NAME}][$(date)][ERR!] Connection was lost!"
sleep ${TIME_FAILED_CONN}
}
LINK () {
if [ ${LOCAL_OPEN} -eq 0 ]; then # Use Reverse SSH Tunneling
	REMOTE_PFWD="-R ${LOCAL_PORT}:127.0.0.1:22"; LOCAL_IP="127.0.0.1"; echo "[${CONNECTION_NAME}][$(date)][INFO] Reverse Conection will be used"
elif [ ${LOCAL_OPEN} -eq 1 ]; then # Use Direct Connection
	REMOTE_PFWD="";	if [ ${LOCAL_IP} = "~" ]; then LOCAL_IP="$(curl ipinfo.io/ip 2> /dev/null)"; fi; fi
echo "[${CONNECTION_NAME}][$(date)][INFO][>>>] Establishing Control..." 
if [ ${DRY} -eq 1 ]; then echo "${CMD_SSH} ${REMOTE_HOST} -l ${REMOTE_USER} -p ${REMOTE_PORT} -i ${KEY} ${COMMON_OPT} ${REMOTE_PFWD} <<"; fi
${CMD_SSH} ${REMOTE_HOST} -l ${REMOTE_USER} -p ${REMOTE_PORT} -i ${KEY} ${COMMON_OPT} ${REMOTE_PFWD} << EOF
	if [ ${1} = 1 ]; then
		if [ ${REMOTE_KILL} -eq 1 ]; then netstat -tlpn | grep ":${REMOTE_FWDPORT} " | sed -n 's@.* \([0-9]*\)/.*@kill \1@p' | sh > /dev/null; fi
		echo "[${CONNECTION_NAME}][$(date)][INFO][<<<] Linked!"; pkill -f "^${CMD_SSH}.*${LOCAL_PFWD_LAST}$" > /dev/null
		if [ ${DRY} -eq 1 ]; then echo "${LOCAL_USER}@${LOCAL_IP}:${LOCAL_PORT} -i ${KEY} ${COMMON_OPT} ${LOCAL_OPT}${REMOTE_PORTPUB} ${LOCAL_PFWD}"
		else ${CMD_SSH} ${LOCAL_IP} -l ${LOCAL_USER} -p ${LOCAL_PORT} -i ${KEY} ${COMMON_OPT} ${LOCAL_OPT}${REMOTE_PORTPUB} ${LOCAL_PFWD}; fi
		echo "[${CONNECTION_NAME}][$(date)][ERR!][<<<] ETOL"
	elif [ ${1} = 2 ]; then
		echo "[${CONNECTION_NAME}][$(date)][INFO][<<<] Dropping...!"
		if [ ${DRY} -eq 1 ]; then pgrep -f "^${CMD_SSH}.*${LOCAL_PFWD_LAST}$"; else pkill -f "^${CMD_SSH}.*${LOCAL_PFWD_LAST}$" > /dev/null; fi; fi
EOF
if [ ${DRY} -eq 1 ]; then exit 0; fi
}
# PARSE INPUT ##########################################################################################################
if [ $# -lt 1 ]; then echo "No Data"; exit 1; fi
REMOTE_PORTPUB=""; DRY=0; LOCAL_OPEN=1; REMOTE_KILL=0; REMOTE_SETUP=0; LOOPCON=1; LOCAL_PFWD="";
while getopts "m:n:h:p:u:H:P:U:D:LSRKX" opt; do 
  case $opt in
  	m) OPER_MODE=${OPTARG};;
	n) CONNECTION_NAME=${OPTARG};;
	h) REMOTE_HOST=${OPTARG};;
	p) REMOTE_PORT=${OPTARG};;
	u) REMOTE_USER=${OPTARG};;
	H) LOCAL_IP=${OPTARG};;
	P) LOCAL_PORT=${OPTARG};;
	U) LOCAL_USER=${OPTARG};;
	D) LOCAL_PFWD="${LOCAL_PFWD}-L ${OPTARG} "; LOCAL_PFWD_LAST=${OPTARG};;
	L) REMOTE_PORTPUB=" -g";;
	S) REMOTE_SETUP=1;;
	R) LOCAL_OPEN=0;;
	K) REMOTE_KILL=1;;
	X) DRY=1;;
    \?) echo "[PEBKAC] WTF is -$OPTARG?, thats not a accepted option, Abort"; USAGE; exit 1;;
    :) echo "[PEBKAC] -$OPTARG requires an argument, Abort"; USAGE; exit 1;;
  esac
done
# POST VAR #############################################################################################################
TIME_FAILED_CONN=2; TIME_FAILED_INET=5; TIMEOUT_VERIFY_INET=15; HOST_VERIFY="https://google.com"
CMD_SSH="ssh"; CMD_SCP="scp"; KEY=lantis.key; SETUP_KEY="$HOME/.ssh/id_rsa"
COMMON_OPT="-C -2 -o BatchMode=yes -o StrictHostKeyChecking=no -o TCPKeepAlive=yes -o ServerAliveInterval=5 -o ConnectTimeout=15 -o LogLevel=Error"
LOCAL_OPT="-N -o CompressionLevel=9 -o ExitOnForwardFailure=yes"
source ./watchdog.lantis.config
# MAIN RUNTIME #########################################################################################################
echo "[${CONNECTION_NAME}][$(date)][INFO] DATA LOADED"
while [ ${LOOPCON} -eq 1 ]; do
		if TEST_INET_VERIFY; then TEST_INET_PASSED
			{ TEST_HOST_VERIFY 
			} || { TEST_HOST_FAILED 
			}; LINK ${OPER_MODE}; if [ ${OPER_MODE} -eq 2 ]; then LOOPCON=0; fi
		else TEST_INET_FAILED; fi; TEST_CONN_FAILED
	done